[Closes #45] Admin/Staff/Volunteer can search allergy/medication/condition data

samau3 opened the pull request:
Closes #45.

This PR creates three protected routes that correspond to querying the allergy, medication, and condition tables of the database. 

- Feedback is provided if no query is provided
- A unique response if a variation of none is used 
- A not found response is provided if the query term isn't found in the respective table

![Allergy Demo](https://github.com/sfbrigade/sf-lifeline/assets/69769431/0d76facb-e4cf-461c-8f68-66405b7736fa)

samau3 commented on server/routes/api/v1/allergies/index.js:
```diff
@@ -0,0 +1,41 @@
+      const allergy = request.query.allergy.trim();
+
+      // prevent empty string from returning all allergies
+      if (!allergy.length) {
+        return reply.send({ message: 'No query provided' });
+      }
```
Since this causes a different response structure, should a different behavior be used when no query parameter is provided?

samau3 commented on server/routes/api/v1/allergies/index.js:
```diff
@@ -0,0 +1,41 @@
+        return reply.send({ message: 'No query provided' });
+      }
+
+      const NO_ALLERGIES = ['none', 'no allergies', 'no known allergies', 'n/a'];
+
+      if (NO_ALLERGIES.includes(allergy.toLowerCase())) {
+        return reply.send({ message: 'No known allergies' });
+      }
```
Similar concern as the situation with no query parameter being provided

francisli replied:
I'm wondering if we need to handle this input... instead, maybe we have a boolean in the Patient record for "no allergies" (and the others) that the user can select with a checkbox, or the user just leaves the section blank

samau3 commented on server/routes/api/v1/allergies/index.js:
```diff
@@ -0,0 +1,41 @@
+        select: { name: true },
+      });
+
+      if (!results.length) {
+        return reply.notFound();
+      }
```
I wasn't sure about how to handle a situation where an unknown but valid allergy/condition/medication is provided, so for now I used a `NOT FOUND` error for this iteration.

francisli reviewed (changes requested):
Hi @samau3 welcome back...!

I was out of town for a few weeks but I'm back and have taken a look here...

This is looking good! A few more changes...

Right now you're only selecting the name to be returned in the results, I think we should return all the attributes of the corresponding object (id, name, type, system, code, etc) so that the caller has everything they might need.

For the pagination, you're manually specifying the take parameter, doing the count, and formatting a string with the results. However, there are some pagination helpers that have since been added into the codebase that you can use instead.

See `server/routes/api/v1/users/list.js` for an example of how these helpers are used. The query helper is in `server/prisma/client.js` and the response helper (which sets the pagination values in headers) is in `server/plugins/pagination.js`. This implementation is based on how the Github does their pagination in their REST APIs (https://docs.github.com/en/rest/using-the-rest-api/using-pagination-in-the-rest-api)

samau3 commented:
Thanks for the feedback @francisli!

> return all the attributes of the corresponding object (id, name, type, system, code, etc) so that the caller has everything they might need.

I've updated all the three routes to return all attributes instead of just the `name`.

> 
> For the pagination, you're manually specifying the take parameter, doing the count, and formatting a string with the results. However, there are some pagination helpers that have since been added into the codebase that you can use 
>
> See `server/routes/api/v1/users/list.js` for an example of how these helpers are used. The query helper is in `server/prisma/client.js` and the response helper (which sets the pagination values in headers) is in `server/plugins/pagination.js`.

I've updated the routes to use the helpers rather than the manual formatting I had before.
